### ISSUE TITLE
Add missing development documentation

### ISSUE THREAD

**Issue by kantord** [LibreLingo/LibreLingo#1152]
- Document where images come from
- Document where sounds come from
- Document previewing courses locally
- Document list of Python packages
- Document setting up Python packages
- Document testing Python packages
- Document Cypress testing
- Document how to write new Gherkin tests

**Comment by jdmaras** [comment]
I was trying to work on documenting where the sounds come from but I was hoping for some clarification on what exactly I was doing.

Am I documenting what site created the sound?

**Comment by pur80a** [comment]
This seems related to these two open issues: 
- https://github.com/LibreLingo/LibreLingo/issues/714
- https://github.com/LibreLingo/LibreLingo/issues/1464
It may be advantageous to document something else, because the sound generation might change and the documentation will need to be rewritten. If I remember correctly, images are similarly stored as sound (at the moment at least). So, writing documentation about how images are handled and pointing out differences to sounds may be something to look into? What do you think?

**Comment by kantord** [comment]
yes, well, those are future plans for where the sound comes from. I think those should be in the documentation and the issues should be linked, because a lot of people are interested in this topic,

the current sound that exists is fetched by a TTS engine through an API, and that is also very important to be documented 

I think the best place to start research could be this closed issue, that is already completed that the pull request where it was fixed is also linked there https://github.com/LibreLingo/LibreLingo/issues/1473

let me know if you need more help regarding this or anything

**Comment by Skivling** [comment]
@kantord 
The existing development documentation does not work with the latest versions of the dependencies, such as yarn. The suggested commands do not exist, `yarn web dev` is now `yarn run web-dev` I believe. When going into the test language, it just redirects to a Hello World page with nothing on it. 
It is also very unclear as to how to run the yaml loader and exporter on a yaml course, with many things just not working or referencing outdated things that don't work anymore. The shell scripts fail and return various errors as well.

Can you please update the documentation and make these testing tools work? Also document how to convert a yaml course to json just using the python tools (which haven't been updated for 2 years??). 

I also tried the TDD python thing and did exactly everything it told me and got errors not accounted in it.

**Comment by Skivling** [comment]
Hang on. Does the yaml export to json just the course content, or is it formatted for how the web app likes it? In a different app (not python based), should I instead be parsing the yaml directly? I will probably read the python yaml-loader and figure out what it does.

**Comment by pablomartan** [comment]
@kantord @Skivling there are a number of changes needed, indeed. For example, it seems like poetry has been replaced by pdm, and also NodeJS v.20 seems to be working just fine. The thing is that, on the main branch, the docs don't include a file for setting up a devenv. I am trying to setup a nix flakes based devenv to ease up that process for myself, and including it for everyone to use would make it much easier, although it would require updating the docs nonetheless. I would be glad to help in any way needed, and wouldn't mind taking charge of it, keeping in mind it could take me a while to figure out all the bells and whistles.

EDIT: I was just checking, and the [docs on contributing](https://github.com/kantord/LibreLingo/blob/main/docs/CONTRIBUTING.md) say the app is written on svelte, but the app on main branch is using react and next...

**Comment by kantord** [comment]
> Hang on. Does the yaml export to json just the course content, or is it formatted for how the web app likes it? In a different app (not python based), should I instead be parsing the yaml directly? I will probably read the python yaml-loader and figure out what it does.

yeah, the JSON export just creates JSON files that the app can consume to display content. building it into a static html page is basically a separate app altogether (although there are other apps out there built with those files, and in theory you can also build an app directly using python-yaml loader)

> EDIT: I was just checking, and the [docs on contributing](https://github.com/kantord/LibreLingo/blob/main/docs/CONTRIBUTING.md) say the app is written on svelte, but the app on main branch is using react and next...

yeah, the web app is being rewritten, as the svelte app is in a total deadlock in terms of dependencies, cannot be fixed, updated or even deployed in it's current stage, and making it work again is as good as a complete rewrite anyways.

**Comment by pablomartan** [comment]
> yeah, the web app is being rewritten, as the svelte app is in a total deadlock in terms of dependencies, cannot be fixed, updated or even deployed in it's current stage, and making it work again is as good as a complete rewrite anyways.

Is there a chance to help on that rewrite? Also, coming back to the main topic of this issue: where can one change the docs? I could submit one or two PRs if I knew where to start :)

**Comment by ggadt** [comment]
> > yeah, the web app is being rewritten, as the svelte app is in a total deadlock in terms of dependencies, cannot be fixed, updated or even deployed in it's current stage, and making it work again is as good as a complete rewrite anyways.
> 
> Is there a chance to help on that rewrite? Also, coming back to the main topic of this issue: where can one change the docs? I could submit one or two PRs if I knew where to start :)

@pablomartan I assume that the source code can be found on the /apps/librelingo-web/* folder 😄 
I'm working on this project when I've free time, if you want let's make up a team so we can collaborate 😄

**Comment by ggadt** [comment]
> @kantord The existing development documentation does not work with the latest versions of the dependencies, such as yarn. The suggested commands do not exist, `yarn web dev` is now `yarn run web-dev` I believe. When going into the test language, it just redirects to a Hello World page with nothing on it. It is also very unclear as to how to run the yaml loader and exporter on a yaml course, with many things just not working or referencing outdated things that don't work anymore. The shell scripts fail and return various errors as well.
> 
> Can you please update the documentation and make these testing tools work? Also document how to convert a yaml course to json just using the python tools (which haven't been updated for 2 years??).
> 
> I also tried the TDD python thing and did exactly everything it told me and got errors not accounted in it.

@pablomartan Would you like to collab with me to clean some mess, since i succeeded to run in a dockerized environment the python apps and modules?

**Comment by aditya-path** [comment]
Hi @pablomartan could you please share dockerfile and commands to run it?
